File: Senior Yr/Fall/cs380/Assignments/Assignment3/src/SlidingBrickPuzzle/A_star.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package SlidingBrickPuzzle;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.PriorityQueue;

/**
 *
 * @author sunnypatel
 */
public class A_star {

    public Node root;
    public LinkedList<Node> tree;
    Comparator<Node> openedComparator = new OpenedComparator();
    public PriorityQueue<Node> opened;
    //public HashMap<Double,LinkedList<Node>> closed;
    public ArrayList<Node> closed;
    
    public A_star(Puzzle initState) {
        tree = new LinkedList<Node>();
        opened = new PriorityQueue<Node>(10, openedComparator);
        //closed = new HashMap<Double,LinkedList<Node>>();
        closed = new ArrayList<Node>();
        // create root node
        this.root = new Node(initState);
        // add root node to tree
        this.tree.addLast(this.root);
    }
    
    public void run(){
        this.AStar_search(root);
    }
    
    public void AStar_search(Node start) {
        start.g = 0;
        start.h = heuristic(start);
        opened.add(start);

        while (!opened.isEmpty()) {
            
            // Print out the open array
            
            
      /*
            for( Node open : opened){
                if(open.move != null){
                open.move.printMove();
               System.out.print("  |  ");
                }
            }
            System.out.println();
            for( Node open : opened){
                if(open.move != null){
                System.out.print("  f()=  "+(open.g+open.h) + "  ");
                }
            }
            System.out.println();
     */
            
            
            
            
           // Node n = this.removeLowestF(opened);
            Node n = opened.poll();
      /*
            if(n.move != null){
                System.out.print("Picked:");
                n.move.printMove();
                System.out.println();
            }
       */
            if (this.goalTest(n)) {
                // return path to n
                //System.out.println("Solved!");
                
                this.printPath(n);
                n.state.printDimensions();
                n.state.printBoard();
                return;
            }
            // The linkedList with all nodes of the same h
            // adding nodes to closed list, hopefully...
         /*   double distFromZero = this.distFromZero(n);
            if(closed.get(distFromZero) != null){
                closed.get(distFromZero).addFirst(n);
            } else{
                LinkedList<Node> tmp = new LinkedList<Node>();
                tmp.addFirst(n);
                closed.put(distFromZero, tmp);
            }
            */
           closed.add(n);
            
            ArrayList<Move> allMoves = n.state.calculateAllMoves(n.state);
            
            
       /*
           System.out.println();
            System.out.println("************");
            n.state.printBoard();
           // System.out.println("g= " +n.g);
           // System.out.println("h= " +n.h);
            
            System.out.println("List all possible moves:");
            System.out.println("l=left, r=right, u=up, b=bottom");
       
            for (Move move : allMoves){
                System.out.println("allMoves="+move.getPiece()+" to "+ move.getDirection());
            }
            
       
            System.out.println("************");
            System.out.println();
        */
            
            
            // create children
            for (Move move : allMoves) {
                Node newChild = new Node(n);

                newChild.g = n.g + 1;
                
                newChild.move = move;
    //            System.out.println("Before:");
    //           n.state.printBoard();
                newChild.state = n.state.applyMoveSeparate(n.state, move);
    //            System.out.println("After:");
    //            newChild.state.printBoard();
                
                newChild.h = this.heuristic(newChild);
                
                // add as child only if not in closed already
                if (!checkIfClosed(newChild)) {
                    n.state.normalize();
                    n.addChild(newChild);
                    opened.add(newChild);
    /*                System.out.print("Made move: ");
                    move.printMove();
                   System.out.println();
   */                 
                    
                } else {
     /*         
                    System.out.print("Move was already made: ");
                    move.printMove();
                   System.out.println();
     */              
                }
                //System.out.println("done");
            }
            
            // temp ish
            int openSize = opened.size();
            if(openSize%1000 == 0){
          //      System.out.println("Opened size= " + openSize);
          //      System.out.println("Closed size= " +closed.size());
                /*for(Node closed : closed){
                    System.out.println("******************");
                    closed.state.printBoard();
                    System.out.println("******************");
                }*/
            }
        }
    }

    public void printPath(Node n){
        ArrayList<Move> successFullMoves = new ArrayList<Move>();
        //Node parent = n;
        
        while(n.parent != null){
            successFullMoves.add(0,n.move);
            n = n.parent;
        }
        for(Move m : successFullMoves){
            m.printMove();
            System.out.println();
        }
    }
    
    public boolean goalTest(Node a) {
        return a.state.checkPuzzle();
    }

    /**
     * Check if child node exists in closedNode already
     *
     * @param child
     * @param closedNode
     * @return
     */
    private boolean checkIfClosed(Node child) {
        if(child.move == null){
           // System.out.println("Child move is null");
            return false;
        }
      /*  LinkedList<Node> tmpBucket = closed.get(child.h);
        if(tmpBucket == null)
            return false;
        
        for(Node bucketNode : tmpBucket){
            if(bucketNode.equals(child))
                return true;
        }
        return false;
        */
        
        for (Node closeNode : this.closed) {
            if(closeNode.move == null){
               // System.out.println("Node in closed is move=null");
                //return false;
            } 
            else if (closeNode.move != null && child.move != null){
               // System.out.println("Both Nodes have moves**");
            }
            if (closeNode.equals(child)) {
               return true;
            }
        }
        return false;
        
        
    }

    private Node removeLowestF(ArrayList<Node> opened) {
        Node lowestF = opened.get(0);
        int indexOfLowestF = 0;
        for(int i=0; i< opened.size();i++){
            Node a = opened.get(i);
            if ((a.g + a.h) < (lowestF.g + lowestF.h)) {
                lowestF = a;
                indexOfLowestF = i;
            }           
        }
        // remove lowest F node from opened
        opened.remove(indexOfLowestF);
        return lowestF;
    }

    /**
     * Pretty ugly heuristic because it will run x*y time on a x*y board.
     *
     * @param a node
     * @return shortest distance to goal
     */
    public double heuristic(Node a) {
        ArrayList<Position> master = new ArrayList<Position>();
        ArrayList<Position> goal = new ArrayList<Position>();

        for (int w = 0; w < a.state.getW(); w++) {
            for (int h = 0; h < a.state.getH(); h++) {
                int piece = a.state.getPiece(w, h);
                if (piece == 2) {
                    // add you're position to masterBlock arrayList<Posotion>
                    master.add(new Position(w, h));

                } else if (piece == -1) {
                    // you're the goal block add yourself to the goal location
                    goal.add(new Position(w, h));
                }
            }
        }
        // Here we know that the master block is on the goal
        if(goal.isEmpty())
            return 0;
        
        
        // TODO now that we know where the goal block and master block,
        //      calculate the difference between them
        // Hold min distance
        double min = -1;
        for (Position mBlock : master) {
            for (Position gBlock : goal) {
                // Use the distance formula
                int tmpX = mBlock.getX() - gBlock.getX();
                int tmpY = mBlock.getY() - gBlock.getY();
                /*
                int tmp = (tmpX * tmpX) + (tmpY * tmpY);
                double dist = Math.sqrt(tmp);
                */
                double dist = tmpX + tmpY;
                if (min == -1) {
                    min = dist;
                } else if (dist < min) {
                    min = dist;
                }
                
                
                
            }
        }

        
        return min;
    }
    
    
        public double distFromZero(Node a) {
        ArrayList<Position> master = new ArrayList<Position>();
        ArrayList<Position> zero = new ArrayList<Position>();

        for (int w = 0; w < a.state.getW(); w++) {
            for (int h = 0; h < a.state.getH(); h++) {
                int piece = a.state.getPiece(w, h);
                if (piece == 2) {
                    // add you're position to masterBlock arrayList<Posotion>
                    master.add(new Position(w, h));

                } else if (piece == 0) {
                    // you're the zero block add yourself to the zero location
                    zero.add(new Position(w, h));
                }
            }
        }
        // Here we know that the master block is on the zero
        if(zero.isEmpty())
            return 0;
        
        
        // TODO now that we know where the zero block and master block,
        //      calculate the difference between them
        // Hold min distance
        double max = -1;
        for (Position mBlock : master) {
            for (Position gBlock : zero) {
                // Use the distance formula
                int tmpX = mBlock.getX() - gBlock.getX();
                int tmpY = mBlock.getY() - gBlock.getY();
                /*
                int tmp = (tmpX * tmpX) + (tmpY * tmpY);
                double dist = Math.sqrt(tmp);
                */
                double dist = tmpX + tmpY;
                if (max == -1) {
                    max = dist;
                } else if (dist > max) {
                    max = dist;
                }
                
                
                
            }
        }

        
        return max;
    }
}

File: Senior Yr/Fall/cs457/PA2-tests/svp28-PA2/Astar.cpp
#include "Astar.h"



Astar::Astar(){

}

Astar::Astar(Puzzle initState){
	root.g =0;
	root.h = 0;
	root.state = initState;
	root.parent = NULL;
	//root.state.printBoard();
}

Astar::~Astar(){

}
vector<Move> Astar::printPath(Node* n){
	vector<Move> successMoves;
	vector<Move>::iterator it = successMoves.begin();
	int numMoves = 0;
	n->state.printBoard();

	for(int i =0; i < n->movesSoFar.size(); i++){
		n->movesSoFar[i].printMove();
		cout << endl;

		numMoves++;
	}
	cout << "# of moves " << numMoves << endl;
	return n->movesSoFar;
	
}

void Astar::printPathRec(Node* n){
	if(n->parent != NULL){
		printPathRec(n->parent);
	}
	n->move.printMove();
}

vector<Move> Astar::Astar_search(){
	root.g = 0;
	root.h = heuristic(&root);

	opened.push(root);

	while(!opened.empty()){
		
		Node n = opened.top();
		//Applyingcout << "openning --" << (n.h + n.g)<< endl;
		//cout << "Astar_search while loop";
		//n.state.printBoard();
	cout << "888888 REMOVEING 888888888 | open.size()=" << opened.size();
		opened.pop();
		cout << "  |   opened.size() after pop() = " << opened.size() << endl;
		cout << " ** Node n popped from open **" << endl;
		n.state.printBoard();
		
		cout << " n.f = " << (n.h + n.g) << endl;
	
		if(goalTest(&n)){
			cout << "goal reached!" << endl;
			return printPath(&n);
			
		}
		vector<Node>::iterator it = closed.begin();
		closed.insert(it, n);

		vector<Move> allMoves = n.state.calculateMoves();

		for(int moveInx=0; moveInx < allMoves.size(); moveInx++){
			Node newChild(n);

			newChild.g = n.g + 1;
			newChild.depth = n.depth + 1;
			newChild.move = allMoves[moveInx];
			newChild.parent = &n;

			newChild.addMove(allMoves[moveInx]);

			newChild.state = n.state.applyMoveSeparate(allMoves[moveInx]);
			
			newChild.h = heuristic(&newChild);


			if(!checkIfClosed(&newChild)){
			cout << "--- adding child ---" << endl;
				newChild.state.printBoard();
			cout << " f = " << (newChild.h + newChild.g)<< endl;
			cout << endl;
		
				n.addChild(newChild);
				opened.push(newChild);
			}
		}
		cout << "open size : " << opened.size() << endl;
		cout << "closed size : " << closed.size() << endl;
		/*if(opened.size() == 100){
		//	cout << "open size : " << opened.size() << endl;
			cout << "********** printing open ****** " << endl;
			printOpen();	
		}
		*/
	}
}

void Astar::printOpen(){
	while(!opened.empty()){
		Node x = opened.top();
		cout << " " << x.h + x.g;
		opened.pop();
	}
}

bool Astar::checkIfClosed(Node* child){
	

	for(int i=0; i<closed.size(); i++){
		if(closed[i].equals(child))
			return true;
	}
	return false;
}

bool Astar::goalTest(Node* n){
	vector<int> tmpBoard = n->state.board;

/*	for(int i=0; i< tmpBoard.size(); i++){
		if( i == 15){
			if(tmpBoard[15] != 0)
				return false;
		}
		if(tmpBoard[i] != i+1)
			return false;


	}
*/
	sleep(100);		
	if(heuristic(n) == 0)
		return true;
	else
		return false;
	
	cout << "IN GOALTEST = TRUE" << endl;
	return true;
	
}

int Astar::heuristic(Node* a){
	int h = 0;
	Puzzle* tmp = &a->state;

	for(int i=0; i < tmp->getBoard().size() ; i++){
		h = h + distanceFromGoal(tmp->findAt(i), i, tmp->getK());
	}
		
	return h;
}

int Astar::distanceFromGoal(int piece, int pos, int k){

	int x1=0;
	int x2=0;
	int y1=0;
	int y2=0;

	//piece = piece - 1;
	if(piece == 0) {
		x1 = k-1;
		y1 = k-1;
	} else {
	  // X1 and Y1 is the position of the piece in the solved puzzle board
		x1 = piece % k;
		y1 = ((piece - x1) / k) + 1;  // which 'row' piece is on in the solved puzzle
	}


	// X2 and Y2 is the position of piece in the current puzzle state
	x2 = pos % k;
	y2 = ((pos - x2) / k) + 1;

	// use the variation of distance formula
	int dist = abs(x1 - x2) + abs(y1 - y2);

	// if i calulated this correctly, it should be the manhattan distance 
	return dist;

}
